src/ctap2/client_data.rs

```diff
@@ -273,27 +275,41 @@ impl<'de> Deserialize<'de> for ClientDataHash {
     }
 }
 
-impl CollectedClientData {
-    pub fn hash(&self) -> json::Result<ClientDataHash> {
+#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
```

Comment:
If we derive `Serialize` but also include the serialized data as a field, won't this repeat the data twice in the serialization? Do we even need this wrapper to be `Serialize`/`Deserialize`?

Reply:
Okay, so there is a very critical reason for this. The hashes of the collected client data are over the *bytes* of the serialised json. This means that the serialised bytes are what need to be sent to the webauthn rp. Previously, this library only returned the collected client data and the *caller* has to then re-serialise the content to bytes, and rely on stable and perfect identical serialisation else the serialised bytes would differ, and that would cause the webauthn rp to fail the ceremony. 

So now the library returns the "structure" of collected client data, and it's serialised content as it was when the authenticator hashed it's content.

Now, the wrapper structure being serialise over this I'm not sure about, since the serialised content is now self contained, but that's probably something @msirringhaus is better able to answer.

Reply:
@rmartinho is right. It makes no sense to have `Serialize`/`Deserialize` for this struct. Serialization of the actual client data happens inside `new()`. And if one wants to serialize `CollectedClientData` again (for whatever reason), the unserialized object is also a member of the wrapper-struct. But serializing the wrapper-struct itself makes no sense.
I'm removing both.